File: python/scenario/voice/messages.py
"""
Audio message helpers.

Encodes AudioChunks into OpenAI-compatible multimodal messages (input_audio
content parts for user role, assistant messages with a transcript + audio
attachment for assistant role) and extracts them back out.

This is the glue between AudioChunk (internal format) and
ChatCompletionMessageParam (the SDK's existing message bus).
"""

from __future__ import annotations

import base64
from typing import Any, Optional, cast

from openai.types.chat import ChatCompletionMessageParam

from .audio_chunk import AudioChunk, PCM16_SAMPLE_RATE


def _pcm16_to_wav_bytes(pcm: bytes) -> bytes:
    """Wrap raw PCM16 mono bytes at 24kHz in a minimal WAV container."""
    from io import BytesIO
    import wave

    buf = BytesIO()
    with wave.open(buf, "wb") as w:
        w.setnchannels(1)
        w.setsampwidth(2)
        w.setframerate(PCM16_SAMPLE_RATE)
        w.writeframes(pcm)
    return buf.getvalue()


def _wav_bytes_to_pcm16(wav: bytes) -> bytes:
    """Extract raw PCM16 frames from a WAV byte string (24kHz mono expected)."""
    from io import BytesIO
    import wave

    with wave.open(BytesIO(wav), "rb") as w:
        return w.readframes(w.getnframes())


def create_audio_message(
    chunk: AudioChunk,
    role: str = "user",
) -> ChatCompletionMessageParam:
    """
    Turn an AudioChunk into an OpenAI-compatible message.

    The content is a list with an input_audio part carrying base64-encoded WAV.
    If the chunk has a transcript, it is added as a text part alongside the
    audio — this is what lets the judge's text-only path still read the content.

    Audio travels cleanly in any role (user or assistant) per the locked design
    — there is no forceUserRole workaround.
    """
    wav = _pcm16_to_wav_bytes(chunk.data)
    b64 = base64.b64encode(wav).decode()
    parts: list[dict[str, Any]] = [
        {
            "type": "input_audio",
            "input_audio": {"data": b64, "format": "wav"},
        }
    ]
    if chunk.transcript:
        parts.insert(0, {"type": "text", "text": chunk.transcript})
    return cast(ChatCompletionMessageParam, {"role": role, "content": parts})


def extract_audio(message: ChatCompletionMessageParam) -> Optional[AudioChunk]:
    """
    Pull the first audio chunk out of an OpenAI-format message.

    Returns None if the message has no audio content part. Accepts both
    'input_audio' (OpenAI API convention) and 'audio' (alternate providers).
    """
    content = message.get("content") if isinstance(message, dict) else None
    if not isinstance(content, list):
        return None

    transcript: Optional[str] = None
    for part in content:
        if not isinstance(part, dict):
            continue
        if part.get("type") == "text":
            transcript = part.get("text") or transcript
        if part.get("type") in ("input_audio", "audio"):
            data_obj = part.get("input_audio") or part.get("audio") or {}
            b64 = data_obj.get("data") if isinstance(data_obj, dict) else None
            if not b64:
                continue
            raw = base64.b64decode(b64)
            # We expect WAV by convention. If it's raw PCM, bytes pass through.
            if raw[:4] == b"RIFF":
                pcm = _wav_bytes_to_pcm16(raw)
            else:
                pcm = raw
            return AudioChunk(data=pcm, transcript=transcript)
    return None


def message_has_audio(message: ChatCompletionMessageParam) -> bool:
    """True if the message contains any audio content part."""
    return extract_audio(message) is not None

File: javascript/src/voice/messages.ts
/**
 * Audio message gateway — the SOLE runtime `AudioChunk ↔ ModelMessage`
 * encoder/extractor. The ONE place the in-message audio format is built and
 * parsed (EDR §4.2, Gap #3).
 *
 * Format: the canonical {@link AudioChunk} (raw PCM16 / 24 kHz / mono) is
 * carried as an **AI-SDK `file` part** — `{ type: "file", mediaType:
 * "audio/pcm16", data: <base64> }` — with the transcript (when present) as a
 * preceding `{ type: "text" }` part so the judge's text-only path still
 * reads the content. This matches what `realtime/response-formatter.ts`
 * already emits and what `judge-utils.ts#buildTranscriptFromMessages`
 * already truncates.
 *
 * Why one encoder/extractor: before this, two producers disagreed —
 * `messages.ts` wrapped PCM16 in a WAV container tagged `format:"wav"`,
 * while `adapter.runtime.ts` emitted raw PCM16 tagged `format:"pcm16"`, both
 * under the OpenAI `input_audio` convention. Their paired extractors decoded
 * by tag, so cross-feeding mis-decoded a WAV header as audio samples (the
 * Gap #3 LIVE BUG). Now there is one raw-PCM16 `file`-part encoder and one
 * extractor; OpenAI `input_audio`/`audio` shapes stay at the adapter edge.
 *
 * No STT/TTS, no provider-native shapes, no state.
 */

import { AudioChunk } from "./audio-chunk";
import type {
  AudioFilePart,
  AudioMessage,
  AudioMessageRole,
} from "./messages.types";
import type { ModelMessage, TextPart } from "ai";

/** Media type of the canonical in-message audio part (raw PCM16). */
export const AUDIO_PCM16_MEDIA_TYPE = "audio/pcm16" as const;

// ---------------------------------------------------------------------------
// base64 helpers (Node Buffer when available, browser btoa/atob fallback)
// ---------------------------------------------------------------------------

function toBase64(data: Uint8Array): string {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(data).toString("base64");
  }
  let binary = "";
  for (let i = 0; i < data.length; i++) binary += String.fromCharCode(data[i]!);
  return btoa(binary);
}

function fromBase64(s: string): Uint8Array {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(s, "base64"));
  }
  const binary = atob(s);
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

/**
 * Strip a 44-byte standard PCM WAV header (scanning for the `data` chunk)
 * so a legacy WAV-wrapped payload still decodes to raw PCM16. New payloads
 * are raw PCM16 and skip this entirely (no RIFF magic).
 */
function wavBytesToPcm16(wav: Uint8Array): Uint8Array {
  const view = new DataView(wav.buffer, wav.byteOffset, wav.byteLength);
  for (let i = 12; i < wav.length - 8; i++) {
    if (wav[i] === 0x64 && wav[i + 1] === 0x61 && wav[i + 2] === 0x74 && wav[i + 3] === 0x61) {
      const dataSize = view.getUint32(i + 4, true);
      const start = i + 8;
      return wav.slice(start, start + dataSize);
    }
  }
  return wav.slice(44);
}

/** Decode an audio payload to raw PCM16 — unwraps a WAV container if present. */
function decodePcm16(b64: string): Uint8Array {
  const raw = fromBase64(b64);
  const isWav =
    raw[0] === 0x52 && raw[1] === 0x49 && raw[2] === 0x46 && raw[3] === 0x46; // "RIFF"
  return isWav ? wavBytesToPcm16(raw) : raw;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Build the canonical audio {@link ModelMessage} from an {@link AudioChunk}.
 *
 * The content is `[textPart?, fileAudioPart]`: a raw-PCM16 AI-SDK `file`
 * part (`mediaType: "audio/pcm16"`), preceded by a transcript text part when
 * the chunk carries one. Audio travels cleanly in any role — no
 * `forceUserRole`.
 *
 * @param chunk - The audio chunk to wrap.
 * @param role  - The message role. Defaults to `"user"`.
 */
export function createAudioMessage(
  chunk: AudioChunk,
  role: AudioMessageRole = "user",
): AudioMessage {
  const audioPart: AudioFilePart = {
    type: "file",
    mediaType: AUDIO_PCM16_MEDIA_TYPE,
    data: toBase64(chunk.data),
  };

  const content: Array<TextPart | AudioFilePart> = [];
  if (chunk.transcript) {
    content.push({ type: "text", text: chunk.transcript });
  }
  content.push(audioPart);

  // The AI-SDK ModelMessage role/content unions are stricter per-role than
  // our "any role carries audio" model; the file+text part array is valid at
  // runtime for user/assistant. Cast keeps the public return type ModelMessage.
  return { role, content } as unknown as AudioMessage;
}

function getContentParts(message: unknown): Array<Record<string, unknown>> | null {
  if (!message || typeof message !== "object") return null;
  const content = (message as Record<string, unknown>)["content"];
  if (!Array.isArray(content)) return null;
  return content as Array<Record<string, unknown>>;
}

/**
 * Pull the first audio chunk out of a {@link ModelMessage}-shaped object.
 *
 * Recognizes the canonical AI-SDK `file` part (`mediaType: "audio/*"`), and
 * — for robustness at the transition / adapter edge — the legacy OpenAI
 * `input_audio` and alternate `audio` conventions. Picks up a transcript
 * from a sibling `text` part. Returns `null` if no audio part is present.
 */
export function extractAudio(message: unknown): AudioChunk | null {
  const content = getContentParts(message);
  if (!content) return null;

  let transcript: string | undefined;

  for (const p of content) {
    if (!p || typeof p !== "object") continue;

    if (p["type"] === "text" && typeof p["text"] === "string") {
      transcript = (p["text"] as string) || transcript;
      continue;
    }

    // Canonical: AI-SDK file part with an audio media type.
    if (
      p["type"] === "file" &&
      typeof p["mediaType"] === "string" &&
      (p["mediaType"] as string).startsWith("audio/")
    ) {
      const data = p["data"];
      if (typeof data === "string") {
        return new AudioChunk({ data: decodePcm16(data), transcript });
      }
      if (data instanceof Uint8Array) {
        return new AudioChunk({ data, transcript });
      }
      continue;
    }

    // Legacy adapter-edge shapes (OpenAI input_audio / alternate audio).
    if (p["type"] === "input_audio" || p["type"] === "audio") {
      const dataObj =
        (p["input_audio"] as Record<string, unknown> | undefined) ??
        (p["audio"] as Record<string, unknown> | undefined) ??
        {};
      const b64 = typeof dataObj["data"] === "string" ? (dataObj["data"] as string) : null;
      if (!b64) continue;
      const tx =
        typeof dataObj["transcript"] === "string"
          ? (dataObj["transcript"] as string)
          : transcript;
      return new AudioChunk({ data: decodePcm16(b64), transcript: tx });
    }
  }

  return null;
}

/** Returns `true` if the message contains any audio content part. */
export function messageHasAudio(message: unknown): boolean {
  return extractAudio(message) !== null;
}

/** Returns `true` if the message carries the canonical audio `file` part. */
export function hasAudio(message: unknown): boolean {
  const content = getContentParts(message);
  if (!content) return false;
  return content.some(
    (p) =>
      p &&
      typeof p === "object" &&
      p["type"] === "file" &&
      typeof p["mediaType"] === "string" &&
      (p["mediaType"] as string).startsWith("audio/"),
  );
}

/** Extract the transcript text from a message's leading text part, if any. */
export function extractTranscript(message: unknown): string | undefined {
  const content = getContentParts(message);
  if (!content) return undefined;
  for (const p of content) {
    if (p && typeof p === "object" && p["type"] === "text" && typeof p["text"] === "string") {
      return p["text"] as string;
    }
  }
  return undefined;
}
